src/QuiqChatClient.js

```diff
@@ -271,17 +300,31 @@ class QuiqChatClient {
     events: Array<Event> = [],
     sendNewMessageCallback: boolean = true,
   ): void => {
-    const newMessages = differenceBy(messages, this.textMessages, 'id');
-    const newEvents = differenceBy(events, this.events, 'id');
+    const newMessages: Array<TextMessage> = differenceBy(messages, this.textMessages, 'id');
+    const newEvents: Array<Event> = differenceBy(events, this.events, 'id');
 
-    const sortedMessages = sortByTimestamp(this.textMessages.concat(newMessages));
-    const sortedEvents = sortByTimestamp(this.events.concat(newEvents));
+    // Apparently, it's possible (though not common) to receive duplicate messages in transcript response.
+    // We need to take union of new and current messages to account for this
 
-    this.textMessages = sortedMessages;
-    this.events = sortedEvents;
+    // If we found new messages, sort them, update cached textMessages, and send callback
+    if (newMessages.length) {
+      this.textMessages = sortByTimestamp(unionBy(this.textMessages, newMessages, 'id'));
 
-    if (newMessages.length && this.callbacks.onNewMessages && sendNewMessageCallback) {
-      this.callbacks.onNewMessages(this.textMessages);
+      if (this.callbacks.onNewMessages && sendNewMessageCallback) {
+        this.callbacks.onNewMessages(this.textMessages);
+      }
+    }
+
+    // If we found new events, sort them, update cached events, and check if a new registration event was received. Fire callback if so.
+    if (newEvents.length) {
+      this.events = sortByTimestamp(unionBy(this.events, newEvents, 'id'));
+
+      if (newEvents.find(e => e.type === MessageTypes.REGISTER)) {
```

Comment:
nitpick, would be better to combine these and have the onRegistration check be first to short circuit the find.

```javascript
if(this.callbacks.onRegistration && newEvents.find(e => e.type === MessageTypes.REGISTER)) {
  ...
```